docs/design/2020-08-08-recursive-common-table-expression.md

```diff
@@ -0,0 +1,108 @@
+
+# Proposal: Recursive Common Table Expression
+
+- Author(s):     [Pingyu](https://github.com/pingyu) (Ping Yu)
+- Last updated:  2020-08-18
+- Discussion at: https://github.com/pingcap/tidb/issues/17472
+
+
+## Abstract
+
+This proposal proposes to support the __Recursive Common Table Expression (Recursive CTE)__.
+
+
+## Background
+
+From MySQL User Document, [13.2.15 WITH (Common Table Expressions)](https://dev.mysql.com/doc/refman/8.0/en/with.html):
+> A common table expression (CTE) is a named temporary result set that exists within the scope of a single statement and that can be referred to later within that statement, possibly multiple times.
+
+E.g.
+```sql
+WITH RECURSIVE cte (n) AS
+(
+  SELECT 1
+  UNION ALL
+  SELECT n + 1 FROM cte WHERE n < 5
+)
+SELECT * FROM cte;
+```
+
+Result:
+```sql
++------+
+| n    |
++------+
+|    1 |
+|    2 |
+|    3 |
+|    4 |
+|    5 |
++------+
+```
+
+## Proposal
+
+### Seed part & Recursive part
+Recursive CTE has two parts, seed part and recursive part, separated by `UNION ALL` or `UNION [DISTINCT]`.
```

Comment:
It's better to introduce the syntax of recursive CTE beforing explaining what are `seed`, `union all`, and `recursive` parts.